{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#read csv file into pandas dataframe\n",
    "df1 = pd.read_csv('sequencing_data_biochem1.csv') \n",
    "\n",
    "#replace 0's with NaN so that cycles with no reads in all 4 columns will return NaN by idxmax() function\n",
    "df1.replace(to_replace={0.000: np.nan}, inplace=True)\n",
    "\n",
    "#select columns for second cycle of biochem1 and rename columns to match format of call_1\n",
    "second_cycle = df1[['A_2','C_2','G_2','T_2']].rename(columns={'A_2':'A', 'C_2':'C', 'G_2':'G', 'T_2':'T'}) \n",
    "#return column name of max value of each cycle, NaN if 0\n",
    "df1['call_2']=second_cycle.idxmax(axis=1) \n",
    "\n",
    "\n",
    "#replace NaN in base call column with N to indicate no reads\n",
    "df1['call_2'].replace(to_replace={np.nan: 'N'}, inplace=True) \n",
    "#convert NaN back to 0.000 in rest of dataframe\n",
    "df1.fillna(0, inplace=True) \n",
    "#with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "    #print(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compare ref_1 to call_1 and record 1 for every mismatch or 0 if the nt readings match\n",
    "df1.insert(6, 'mismatch_1', np.where(df1['ref_1'] != df1['call_1'], 1, 0)) \n",
    "#compare ref_2 to call_2 and record 1 for every mismatch or 0 if the nt readings match\n",
    "df1['mismatch_2'] = np.where(df1['ref_2'] != df1['call_2'], 1, 0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.9 6.5 5.2\n"
     ]
    }
   ],
   "source": [
    "#calculate error for both cycles and total error for biochem1\n",
    "#done by summing 1's in mismatch column and dividing by total reads and multiplying by 100\n",
    "\n",
    "biochem1_error1 = (sum(df1['mismatch_1']))/len(df1['ref_1'])*100 \n",
    "biochem1_error2 = (sum(df1['mismatch_2']))/len(df1['ref_1'])*100\n",
    "biochem1_totalerror = (sum(df1['mismatch_1']) + sum(df1['mismatch_2']))/(2*len(df1['ref_1']))*100\n",
    "\n",
    "print(biochem1_error1, biochem1_error2, biochem1_totalerror)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_1</th>\n",
       "      <th>A_1</th>\n",
       "      <th>C_1</th>\n",
       "      <th>G_1</th>\n",
       "      <th>T_1</th>\n",
       "      <th>call_1</th>\n",
       "      <th>ref_2</th>\n",
       "      <th>A_2</th>\n",
       "      <th>C_2</th>\n",
       "      <th>G_2</th>\n",
       "      <th>T_2</th>\n",
       "      <th>call_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>0.168</td>\n",
       "      <td>0.232</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.071</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.122</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.244</td>\n",
       "      <td>0.037</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.285</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.155</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>0.321</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.196</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.060</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.152</td>\n",
       "      <td>0.049</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>0.170</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.146</td>\n",
       "      <td>0.088</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.335</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>0.129</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.302</td>\n",
       "      <td>0.056</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.122</td>\n",
       "      <td>0.184</td>\n",
       "      <td>0.061</td>\n",
       "      <td>0.201</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>G</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.104</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>C</td>\n",
       "      <td>0.398</td>\n",
       "      <td>0.186</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.025</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.212</td>\n",
       "      <td>0.128</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.030</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>A</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.089</td>\n",
       "      <td>0.186</td>\n",
       "      <td>0.350</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.035</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.165</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>G</td>\n",
       "      <td>0.168</td>\n",
       "      <td>0.297</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.079</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.090</td>\n",
       "      <td>0.135</td>\n",
       "      <td>0.178</td>\n",
       "      <td>0.235</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>A</td>\n",
       "      <td>0.144</td>\n",
       "      <td>0.191</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.358</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.192</td>\n",
       "      <td>0.231</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ref_1    A_1    C_1    G_1    T_1 call_1 ref_2    A_2    C_2    G_2  \\\n",
       "0       G  0.168  0.232  0.028  0.071      G     T  0.122  0.004  0.244   \n",
       "1       A  0.099  0.064  0.004  0.285      A     A  0.027  0.086  0.100   \n",
       "2       C  0.321  0.148  0.070  0.196      C     C  0.060  0.084  0.152   \n",
       "3       G  0.170  0.360  0.146  0.088      G     A  0.165  0.111  0.004   \n",
       "4       T  0.129  0.111  0.302  0.056      T     A  0.122  0.184  0.061   \n",
       "..    ...    ...    ...    ...    ...    ...   ...    ...    ...    ...   \n",
       "995     G  0.000  0.000  0.000  0.000      N     G  0.079  0.239  0.050   \n",
       "996     C  0.398  0.186  0.175  0.025      C     C  0.212  0.128  0.024   \n",
       "997     A  0.086  0.089  0.186  0.350      A     G  0.035  0.330  0.021   \n",
       "998     G  0.168  0.297  0.157  0.079      G     A  0.090  0.135  0.178   \n",
       "999     A  0.144  0.191  0.024  0.358      A     A  0.109  0.040  0.192   \n",
       "\n",
       "       T_2 call_2  \n",
       "0    0.037      T  \n",
       "1    0.155      A  \n",
       "2    0.049      T  \n",
       "3    0.335      A  \n",
       "4    0.201      A  \n",
       "..     ...    ...  \n",
       "995  0.104      G  \n",
       "996  0.030      C  \n",
       "997  0.165      G  \n",
       "998  0.235      A  \n",
       "999  0.231      A  \n",
       "\n",
       "[1000 rows x 12 columns]"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Biochem 2\n",
    "\n",
    "df2 = pd.read_csv('sequencing_data_biochem2.csv')\n",
    "\n",
    "df2.replace(to_replace = {0.000: np.nan}, inplace=True)\n",
    "second_cycle1 = df2[['A_2','C_2','G_2','T_2']].rename(columns={'A_2':'A', 'C_2':'C', 'G_2':'G', 'T_2':'T'})\n",
    "df2['call_2'] = second_cycle1.idxmax(axis=1)\n",
    "\n",
    "#replace NaN with N\n",
    "df2['call_2'].replace(to_replace = {np.nan: 'N'}, inplace=True)\n",
    "#convert NaN back to 0.000\n",
    "df2.fillna(0, inplace=True) \n",
    "\n",
    "#create dictionary to compensate for software error\n",
    "mydict = {'C':'G', 'T':'A', 'A': 'C', 'G':'T'}\n",
    "\n",
    "#replace values in both call columns with correct reading\n",
    "df2['call_1'] = df2['call_1'].replace(mydict)\n",
    "df2['call_2'] = df2['call_2'].replace(mydict)\n",
    "#df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.insert(6, 'mismatch_1', np.where(df2['ref_1'] != df2['call_1'], 1, 0))\n",
    "df2['mismatch_2'] = np.where(df2['ref_2'] != df2['call_2'], 1, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.5 4.9 5.2\n"
     ]
    }
   ],
   "source": [
    "#calculate error for both cycles and total error for biochem1\n",
    "#done by summing 1's in mismatch column and dividing by total reads and multiplying by 100\n",
    "\n",
    "biochem_2_error_1 = (sum(df2['mismatch_1']))/len(df2['ref_1'])*100\n",
    "biochem_2_error_2 = (sum(df2['mismatch_2']))/len(df2['ref_1'])*100\n",
    "biochem_2_totalerror = (sum(df2['mismatch_1']) + sum(df2['mismatch_2']))/(2*len(df2['ref_1']))*100\n",
    "\n",
    "print(biochem_2_error_1, biochem_2_error_2, biochem_2_totalerror)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0 2.0 2.0 2.0\n"
     ]
    }
   ],
   "source": [
    "##Further Analysis\n",
    "\n",
    "####Below code tests to see what percentage of mismatches were result of no reads (N)\n",
    "\n",
    "percent_nans1 = (sum(np.where(df1['call_1'] == 'N', 1, 0)))/len(df1['ref_1'])*100\n",
    "percent_nans2 = (sum(np.where(df1['call_2'] == 'N', 1, 0)))/len(df1['ref_2'])*100\n",
    "percent_nans3 = (sum(np.where(df2['call_1'] == 'N', 1, 0)))/len(df2['ref_1'])*100\n",
    "percent_nans4 = (sum(np.where(df2['call_2'] == 'N', 1, 0)))/len(df2['ref_2'])*100\n",
    "\n",
    "print(percent_nans1, percent_nans2 ,percent_nans3 ,percent_nans4) #both cycles of both biochemistries return 2% N's, so this is not a factor that contributes to varying accuracies between two methods of sequencing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  nucleotide    mean_x    mean_y  error_counts  total_counts\n",
      "0          A  0.499485  0.066370            46           509\n",
      "1          C  0.492514  0.082500             4           494\n",
      "2          G  0.497853  0.052400             5           509\n",
      "3          T  0.497550  0.054667             9           488\n"
     ]
    }
   ],
   "source": [
    "###generate summary chart for biochemistry 1\n",
    "##find average signal intensity values for correct base calls\n",
    "\n",
    "#subset the correct reads from cycle 1 of biochem 1\n",
    "cycle1 = df1.iloc[:,1:6].loc[df1['mismatch_1'] == 0]\n",
    "#rename columns to match those of cycle 2\n",
    "cycle1.rename(columns={'A_1':'A', 'C_1':'C', 'G_1':'G', 'T_1':'T', 'call_1': 'call'}, inplace=True)\n",
    "\n",
    "#subset the correct reads from cycle 2 of biochem 1\n",
    "cycle2 = df1.iloc[:,8:13].loc[df1['mismatch_2'] == 0]\n",
    "#renaame columns to match those of cycle 1\n",
    "cycle2.rename(columns={'A_2':'A', 'C_2':'C', 'G_2':'G', 'T_2':'T', 'call_2': 'call'}, inplace=True)\n",
    "\n",
    "#add the second cycle to the bottom of the first cycle\n",
    "biochem1 = cycle1.append(cycle2, ignore_index=True)\n",
    "\n",
    "\n",
    "#find the max signal intensity for all rows, representing the signal intensity for correct reads\n",
    "biochem1['max'] = biochem1.max(axis=1)\n",
    "#remove signal intensity reading columns\n",
    "biochem1.drop(columns=['A', 'C', 'G', 'T'], inplace=True)\n",
    "#find the mean of signal intensities for all four bases\n",
    "biochem1['mean'] = biochem1.groupby(['call']).transform('mean')\n",
    "#remove max signal intensity column, and remove duplicates\n",
    "#only the average max intensity for each base is left\n",
    "biochem1 = biochem1.drop(columns=['max']).drop_duplicates()\n",
    "#sort alphabetically\n",
    "biochem1.sort_values(by=['call'], inplace=True)\n",
    "#rename call column to nucleotide\n",
    "biochem1.rename(columns={'call':'nucleotide'}, inplace=True)\n",
    "\n",
    "\n",
    "##find average signal intensity for incorrect base calls and incorrect base call counts\n",
    "#subset the incorrect reads from cycle 1 of biochem 1\n",
    "error1 = df1.iloc[:,0:5].loc[((df1['mismatch_1'] == 1) & (df1['call_1'] != 'N'))].reset_index(drop=True)\n",
    "error1.rename(columns={'ref_1':'ref','A_1':'A', 'C_1':'C', 'G_1':'G', 'T_1':'T'}, inplace=True)\n",
    "\n",
    "#subset the incorrect reads from cycle 2 of biochem 1\n",
    "error2 = df1.iloc[:,7:12].loc[((df1['mismatch_2'] == 1) & (df1['call_2'] != 'N'))].reset_index(drop=True)\n",
    "error2.rename(columns={'ref_2':'ref', 'A_2':'A', 'C_2':'C', 'G_2':'G', 'T_2':'T'}, inplace=True)\n",
    "\n",
    "#add the second cycle to the bottom of the first cycle\n",
    "biochem1_errors = error1.append(error2, ignore_index=True)\n",
    "\n",
    "#function to return the signal intensity of the correct base call\n",
    "def find_errors(df):\n",
    "    ref = df.iloc[:,0]\n",
    "\n",
    "    my_arr = []\n",
    "    i = 0\n",
    "    for j in ref:\n",
    "        my_arr.append(df.loc[i,j])\n",
    "        i += 1\n",
    "\n",
    "    return(my_arr)\n",
    "\n",
    "#create new column to contain signal intensity of intended base call\n",
    "biochem1_errors['error'] = find_errors(biochem1_errors)\n",
    "#remove signal intensity reading columns\n",
    "biochem1_errors.drop(columns=['A', 'C', 'G', 'T'], inplace=True)\n",
    "#find the mean of signal intensities for all four bases\n",
    "biochem1_errors['mean'] = biochem1_errors.groupby(['ref']).transform('mean')\n",
    "#count the number of errors for each nucleotide\n",
    "biochem1_errors['error_counts'] = biochem1_errors.groupby(['ref'])['error'].transform('count')\n",
    "#remove erroneous signal intensity column, and remove duplicates\n",
    "#only the average max intensity for each base is left\n",
    "biochem1_errors = biochem1_errors.drop(columns=['error']).drop_duplicates()\n",
    "#sort alphabetically\n",
    "biochem1_errors.sort_values(by=['ref'], inplace=True)\n",
    "#rename call column to nucleotide\n",
    "biochem1_errors.rename(columns={'ref':'nucleotide'}, inplace=True)\n",
    "\n",
    "\n",
    "#merge two dataframes together to generate summary chart\n",
    "summary1 = pd.merge(biochem1, biochem1_errors, on='nucleotide')\n",
    "#count the total number of each nucleotides on the reference genome and append the total to summary chart\n",
    "total_counts = df1['ref_1'].value_counts() + df1['ref_2'].value_counts()\n",
    "summary1['total_counts'] = total_counts.values\n",
    "summary1.to_csv('summary1.csv', index=False)\n",
    "print(summary1)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  nucleotide    mean_x    mean_y  error_counts  total_counts\n",
      "0          A  0.297766  0.104200            15           502\n",
      "1          C  0.301624  0.119379            29           475\n",
      "2          G  0.299207  0.110667             9           512\n",
      "3          T  0.295464  0.128182            11           511\n"
     ]
    }
   ],
   "source": [
    "###generate summary chart for biochemistry 2\n",
    "##find average signal intensity values for correct base calls\n",
    "\n",
    "#subset the correct reads from cycle 1 of biochem 2\n",
    "cycle3 = df2.iloc[:,1:6].loc[df2['mismatch_1'] == 0]\n",
    "cycle3.rename(columns={'A_1':'A', 'C_1':'C', 'G_1':'G', 'T_1':'T', 'call_1': 'call'}, inplace=True)\n",
    "\n",
    "#subset the correct reads from cycle 2 of biochem 2\n",
    "cycle4 = df2.iloc[:,8:13].loc[df2['mismatch_2'] == 0]\n",
    "cycle4.rename(columns={'A_2':'A', 'C_2':'C', 'G_2':'G', 'T_2':'T', 'call_2': 'call'}, inplace=True)\n",
    "\n",
    "#append second cycle to bottom of first cycle\n",
    "biochem2 = cycle3.append(cycle4, ignore_index=True)\n",
    "\n",
    "#find the max signal intensity for all rows, representing the signal intensity for correct reads\n",
    "biochem2['max'] = biochem2.max(axis=1)\n",
    "#remove signal intensity reading columns\n",
    "biochem2.drop(columns=['A', 'C', 'G', 'T'], inplace=True)\n",
    "#find the mean of signal intensities for all four bases\n",
    "biochem2['mean'] = biochem2.groupby(['call']).transform('mean')\n",
    "#remove max signal intensity column, and remove duplicates\n",
    "#only the average max intensity for each base is left\n",
    "biochem2 = biochem2.drop(columns=['max']).drop_duplicates()\n",
    "#sort alphabetically\n",
    "biochem2.sort_values(by=['call'], inplace=True)\n",
    "#rename call column to nucleotide\n",
    "biochem2.rename(columns={'call':'nucleotide'}, inplace=True)\n",
    "\n",
    "\n",
    "##find average signal intensity for incorrect base calls and incorrect base call counts\n",
    "#subset the incorrect reads from cycle 1 of biochem 2\n",
    "error3 = df2.iloc[:,0:5].loc[((df2['mismatch_1'] == 1) & (df2['call_1'] != 'N'))].reset_index(drop=True)\n",
    "error3.rename(columns={'ref_1':'ref','A_1':'A', 'C_1':'C', 'G_1':'G', 'T_1':'T'}, inplace=True)\n",
    "\n",
    "#subset the incorrect reads from cycle 2 of biochem 2\n",
    "error4 = df2.iloc[:,7:12].loc[((df2['mismatch_2'] == 1) & (df2['call_2'] != 'N'))].reset_index(drop=True)\n",
    "error4.rename(columns={'ref_2':'ref', 'A_2':'A', 'C_2':'C', 'G_2':'G', 'T_2':'T'}, inplace=True)\n",
    "\n",
    "#add the second cycle to the bottom of the first cycle\n",
    "biochem2_errors = error3.append(error4, ignore_index=True)\n",
    "\n",
    "#create new column to contain signal intensity of intended base call\n",
    "biochem2_errors['error'] = find_errors(biochem2_errors)\n",
    "#remove signal intensity reading columns\n",
    "biochem2_errors.drop(columns=['A', 'C', 'G', 'T'], inplace=True)\n",
    "#find the mean of signal intensities for all four bases\n",
    "biochem2_errors['mean'] = biochem2_errors.groupby(['ref']).transform('mean')\n",
    "#count the number of errors for each nucleotide\n",
    "biochem2_errors['error_counts'] = biochem2_errors.groupby(['ref'])['error'].transform('count')\n",
    "#remove erroneous signal intensity column, and remove duplicates\n",
    "#only the average max intensity for each base is left\n",
    "biochem2_errors = biochem2_errors.drop(columns=['error']).drop_duplicates()\n",
    "#sort alphabetically\n",
    "biochem2_errors.sort_values(by=['ref'], inplace=True)\n",
    "#rename call column to nucleotide\n",
    "biochem2_errors.rename(columns={'ref':'nucleotide'}, inplace=True)\n",
    "\n",
    "\n",
    "#merge two dataframes together to generate summary chart\n",
    "summary2 = pd.merge(biochem2, biochem2_errors, on='nucleotide')\n",
    "#count the total number of each nucleotides on the reference genome and append the total to summary chart\n",
    "total_counts = df2['ref_1'].value_counts() + df2['ref_2'].value_counts()\n",
    "summary2['total_counts'] = total_counts.values\n",
    "summary2.to_csv('summary2.csv', index=False)\n",
    "print(summary2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
